server/topdf/topdf.go

```diff
@@ -0,0 +1,143 @@
+func (t *TOPDF) CheckServerStatus() (running bool, err error) {
+	return t.server.Status()
+}
+
+// GetPDF gets PDF for fileID that belongs userID. user has to have access to the file
+// otherwise ErrUnauthorizedUser is returned.
+func (t *TOPDF) GetPDF(userID, fileID string) (pdf io.ReadCloser, err error) {
+	pdf, err = t.getPDF(userID, fileID)
+	if err != nil {
+		// return an authorization error if we got an err from Plugin's API.
+		if _, ok := err.(*model.AppError); ok {
+			return nil, ErrUnauthorizedUser
+		}
+		return nil, err
+	}
+	return pdf, nil
+}
+
+// getPDF gets PDF for fileID that belongs to userID.
+// notes:
+// - Mattermost's Plugin API does not implement io.Reader while dealing with files but this might
+//   be improved in future since large files can pump memory usage. TOPDF created streams in mind,
+//   this is why we use ioutil.NopCloser and bytes.Reader in the code below, to even with current Plugin API.
+// - Mattermost's Plugin API does not return errors as `error`s but returns them as *model.AppError,
+//   this needs to be improved since it causes issues while dealing with errors. For more info
+//   please see: https://golang.org/doc/faq#nil_error
+//   to workaround this, Plugin errors are normalized with normalizeAppErr().
+func (t *TOPDF) getPDF(userID, fileID string) (pdf io.ReadCloser, err error) {
+	// try to get id of PDF file that possibly generated and cached for fileID before.
+	pid, aerr := t.mapi.KVGet(key(fileID))
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// if there is no PDF file cached, create it, cache and use its content.
+	if len(pid) == 0 {
+		data, err := t.createAndSavePDF(fileID, userID)
+		if err != nil {
+			return nil, err
+		}
+		return ioutil.NopCloser(bytes.NewReader(data)), nil
+	}
+	// we have the PDF version in cache, directly return it back.
+	data, aerr := t.mapi.GetFile(string(pid))
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	return ioutil.NopCloser(bytes.NewReader(data)), nil
+}
+
+// createAndSavePDF creates a PDF version of fileID and caches on Mattermost server and returns
+// the pdf data back.
+func (t *TOPDF) createAndSavePDF(fileID string, userID string) (pdf []byte, err error) {
+	// get file's info.
+	info, aerr := t.mapi.GetFileInfo(fileID)
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// get associated post for the file.
+	post, aerr := t.mapi.GetPost(info.PostId)
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// check if the user has access to the channel where associated post submitted.
+	if _, aerr := t.mapi.GetChannelMember(post.ChannelId, userID); aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// get file's content by fileID.
+	byts, aerr := t.mapi.GetFile(fileID)
+	if aerr != nil {
+		return nil, normalizeAppErr(aerr)
+	}
+	// convert file to PDF by using PDF server.
+	r, err := t.server.Convert(info.Name, info.Extension, bytes.NewReader(byts))
```

Comment:
Would it make sense to leverage your distributed lock here to avoid creating it more than once, as well as to allow other plugin requests to block until the file is created? Can we use the file id as the key?

Reply:
Yes, yes!